src/t2c.c

```diff
@@ -0,0 +1,302 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <llvm-c/Analysis.h>
+#include <llvm-c/BitWriter.h>
+#include <llvm-c/Core.h>
+#include <llvm-c/ExecutionEngine.h>
+#include <llvm-c/Target.h>
+#include <llvm-c/Transforms/PassBuilder.h>
+#include <stdlib.h>
+
+#include "jit.h"
+#include "riscv_private.h"
+
+#define MAX_BLOCKS 8152
+
+struct LLVM_block_map_entry {
+    uint32_t pc;
+    LLVMBasicBlockRef block;
+};
+
+struct LLVM_block_map {
+    uint32_t count;
+    struct LLVM_block_map_entry map[MAX_BLOCKS];
+};
+
+FORCE_INLINE void t2c_block_map_insert(struct LLVM_block_map *map,
+                                       LLVMBasicBlockRef *entry,
+                                       uint32_t pc)
+{
+    struct LLVM_block_map_entry map_entry = {
+        .block = *entry,
+        .pc = pc,
+    };
+    map->map[map->count++] = map_entry;
+    return;
+}
+
+FORCE_INLINE LLVMBasicBlockRef t2c_block_map_search(struct LLVM_block_map *map,
+                                                    uint32_t pc)
+{
+    for (uint32_t i = 0; i < map->count; i++) {
+        if (map->map[i].pc == pc) {
+            return map->map[i].block;
+        }
+    }
+    return NULL;
+}
+
+#define T2C_OP(inst, code)                                                \
+    static void t2c_##inst(                                               \
+        LLVMBuilderRef *builder UNUSED, LLVMTypeRef *param_types UNUSED,  \
```

Comment:
The `LLVMBuilderRef *builder` is for receiving the change of control flow in most `slt`-like RISC-V insturctions.
Can we use `LLVMBuildSelect` alternatively to eliminate the pinter to builder and reduce the potential control flow change?

Reply:
I have tried `LLVMBuildSelect` before the current implementation, but it would cause error. For example, you can try to replace it with `LLVMBuildSelect(*builder, cmp, T2C_LLVM_GEN_STORE_IMM32(*builder, 1, addr_rd), T2C_LLVM_GEN_STORE_IMM32(*builder, 0, addr_rd),"");`, and make the simulation exeute only through T2C path.

Reply:
The error output of running `fcalc.elf` is
```
./build/rv32emu ./build/fcalc.elf 
Test 'Integer' failed; Expected 0.00000000000000000000000000000000000000000011/0, got 0.000000000000000000000000000000000000000000141937561/0
11Integer11.32Real number.32Sub-0-Real-42Negation1.2 + 5Addition3 - 1.4Subtraction10 * 5Multiplication50 / 10Division3 - 2 * 5PrecedencepiPi3 * (1 + 2)Parentheses(1 + 2Unmatched (Performed 0 tests, 0 failures, 0% success rate.
INFinfNANnan0123456789abcdef0123456789ABCDEF(null)0/POSIXCASCIIISO-8859-1cjksinglecjknarrowcjkwideUTF-8UTF8JISEUCJPEUCJPSJISISO8859ISO-8859-CPKOI8CP20866CP21866GEORGIANPSCP101PT154CP102TIS620CP874LC_ALLLANG.LC_COLLATELC_CTYPELC_MONETARYLC_NUMERICLC_TIMELC_MESSAGESInfinityNaNBalloc succeeded/Users/ilg/actions-runners/xpack-dev-tools/_work/riscv-none-elf-gcc-xpack/riscv-none-elf-gcc-xpack/build/darwin-arm64/sources/newlib-4.3.0.20230120/newlib/libc/stdlib/dtoa.c/Users/ilg/actions-runners/xpack-dev-tools/_work/riscv-none-elf-gcc-xpack/riscv-none-elf-gcc-xpack/build/darwin-arm64/sources/newlib-4.3.0.20230120/newlib/libc/stdlib/mprec.c, function: assertion "(null)" failed: file "(null)", line 011(null)
$��,$�������$�������4�������$���$���$���4���$���������������4���
��,
����,
��,
��,
���
   ��,
��,
����$��,
�������,
������  ���     ���     ���     ���     ���     ���     ���     ���     ��,
��,
��,
��,
��,
��,
��,
��,
��,\��h
��,
��,
��,
��P��,
��,
�����,
��,
��,
��`��,
��4��,
��,
��P��,
��,
��,
��,
��,
��,
��,
��,
��,
����������,
��<��,
��h�����(�����,
��`��8
      �����,
��,
```
Therefore, I believe the current implementation is feasible, as the use of `LLVMBuildSelect` results in incorrect behavior.

Reply:
I tried this and it works for me. It seems that `LLVMBuilderSelect` needs immediate value to be its operands.

```c
LLVMValueRef r1 = LLVMConstInt(LLVMInt32Type(), 1, true);
LLVMValueRef r2 = LLVMConstInt(LLVMInt32Type(), 0, true);
LLVMValueRef _ = LLVMBuildSelect(*builder, cmp, r1, r2, "");
LLVMBuildStore(*builder, _, addr_rd);
```

Reply:
Right, it works for me as well. I misunderstood the use of `LLVMBuildSelect`. I have now modified the implementation accordingly.